Allows jenkins analytics to run analytics tasks on AWS resources

pomegranited opened the pull request:
Allows jenkins analytics jobs to ssh to AWS resources and, using ssh key forwarding, clone secure repos.

Also fixes issue with re-building the edx-analytics-pipeline dependencies (detailed in (closed) [edx-analytics-pipeline PR #228](https://github.com/edx/edx-analytics-pipeline/pull/228)).

* Sets JENKINS_HOME to /edx/var/jenkins in `/etc/defaults/jenkins` instead of using the default `/var/lib/jenkins` symlink.  Accessing JENKINS_HOME via the symlink was causing the pyinstrument pip dependency to fail on re-install.  I've left the symlink in place though, because travis-ci failed without it.
* Creates a second credential (in addition to the git credential) for shelling into the AWS service.  This credential is pulled from a file in the configured secure-repo, which is cloned when the seed task is run.
* Adds missing package dependencies required to build the analytics-pipeline on the jenkins server.

**JIRA tickets**: [OLIVE-32](https://openedx.atlassian.net/browse/OLIVE-32), [OC-1512](https://tasks.opencraft.com/browse/OC-1512)

**Discussions**: See [discussion on PR 15](https://github.com/edx-ops/edx-jenkins-job-dsl/pull/15#issuecomment-204265069).

**Dependencies**:

* [edx-analytics-configuration PR #30](https://github.com/edx/edx-analytics-configuration/pull/30) - fixes ansible 1.4.4 incompatibility
* [edx-jenkins-job-dsl PR #19](https://github.com/edx-ops/edx-jenkins-job-dsl/pull/19) - wraps the shell commands with both sshAgent credentials: github, AWS
* [edx-analytics-pipeline PR #232](https://github.com/edx/edx-analytics-pipeline/pull/232) - fixes bug with initial clone of the secure repo
* [edx-jenkins-job-dsl PR #23](https://github.com/edx-ops/edx-jenkins-job-dsl/pull/23) - fixes bug detected during this PR review, and adds unit tests for AnalyticsTasks.parseYaml

**Sandbox URL**: 

http://52.20.136.133:8080 - ping @pomegranited for login credentials.

**Testing instructions**:

1. Install edx/configuration requirements in a virtualenv:

        virtualenv venv
        . venv/bin/activate
        git clone https://git@github.com/edx/configuration.git
        cd configuration
        make requirements

1. Verify that you have access to https://github.com/open-craft/edx-jenkins-job-dsl/
1. Create a YAML file containing the jenkins analytics configuration.  See [jenkins_analytics README](https://github.com/edx/configuration/tree/master/playbooks/roles/jenkins_analytics) and/or [example ansible-jenkins-credentials.yml.zip](https://github.com/edx/edx-analytics-configuration/files/205786/ansible-jenkins-credentials.yml.zip) for details.
1. Build the vagrant instance, using the YAML file created above:

        cd vagrant/base/analytics_jenkins
        export VAGRANT_JENKINS_LOCAL_VARS_FILE=/path/to/your.yml
        vagrant up

   Expected result is that you will get a working jenkins instance that you'll be able to log in to 192.168.33.15:8080 or localhost:8080, using the `jenkins` user, and the password in `JENKINS_ANALYTICS_USER_PASSWORD_PLAIN`.  The AnalyticsSeedJob will have run successfully, and created an AnswerDistributionWorkflow job.

1. Via your AWS GUI, generate an Access Key for use by the vagrant VM.  URL looks something like this: https://console.aws.amazon.com/iam/home?#users/<your username>
1. Shell into your vagrant instance, and grant the jenkins user access to your AWS instance.

        vagrant ssh
        sudo su jenkins
        cd
        virtualenv aws-venv
        . aws-venv/bin/activate
        pip install awscli
        aws configure
        # AWS Access Key ID [None]: <your key>
        # AWS Secret Access Key [None]: <your code>
        # Default region name [None]: <your region>
        # Default output format [None]:

1. Via the Jenkins GUI, Build the AnswerDistributionWorkflow job with the configured parameters.

Expected result is the task will run successfully, and terminate the EMR at the end.  Previously, the analytics task would fail after EMR provisioning, because it was unable to clone the secure repo on the EMR instance.

Re-running the AnswerDistributionWorkflow will produce the same result.  Previously, every other analytics task build failed due to pyinstrument re-install errors.

**Reviewers**
- [x] @jbzdak 
- [x] @mulby 
- [ ] @jibsheet 

**Settings**

* See [jenkins_analytics README](https://github.com/open-craft/configuration/blob/jill/olive-22-jenkins-sshagent/playbooks/roles/jenkins_analytics/README.md) for details on the ansible variables.
* See [ansible-jenkins-credentials.yml.zip](https://github.com/edx/configuration/files/209720/ansible-jenkins-credentials.yml.zip) for example.

edx-webhook commented:
Thanks for the pull request, @pomegranited! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fconfiguration&number=2939

pomegranited commented:
@mulby [add0f077](https://github.com/edx/configuration/pull/2939/commits/add0f077f53d552558b0227c3e38d0d041c8917e) changes the <ANALYTICS_TASK>_EXTRA_VARS from a JSON string to a @path/to/file.yml.  Does this help with your workflow?

jbzdak commented on vagrant/base/analytics_jenkins/Vagrantfile:
```diff
@@ -14,10 +14,18 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
+
   unless ENV['VAGRANT_NO_PORTS']
-    config.vm.network :forwarded_port, guest: 8080, host: 8080  # Jenkins
+    #config.vm.network :forwarded_port, guest: 8080, host: 8080  # Jenkins
```
Commented line.

jbzdak commented on vagrant/base/analytics_jenkins/Vagrantfile:
```diff
@@ -14,10 +14,18 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
 
+  # config.vm.provider :virtualbox do |vb|
+  #   vb.gui = true
+  # end
```
Commented code (feel free to add gui like: ``if ENV['VAGRANT_GUI']``).

pomegranited replied:
Oops! I did not intend to commit that file. Will revert at first
opportunity.

jbzdak commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -179,14 +194,19 @@ The full list of seed job configuration variables is:
-* `ANALYTICS_SCHEDULE_<TASK_NAME>_EXTRA_VARS`: YML config or @file location to
-  override the analytics task parameters.
+* `ANALYTICS_SCHEDULE_<TASK_NAME>_EXTRA_VARS`: YML @file location to
+  override the analytics task parameters.  File locations must be absolute, or
```
"File locations must be absolute or relative" -> "File locations can be either absolute or relative"

jbzdak commented on playbooks/roles/jenkins_analytics/templates/seed_job_template.xml:
```diff
@@ -18,60 +18,66 @@
-                <defaultValue>{{ ANALYTICS_SCHEDULE_INSERT_TO_MYSQL_COURSE_ENROLL_BY_COUNTRY_EXTRA_VARS | default('{}') | to_nice_json }}</defaultValue>
-                <description>Set default values for the InsertToMysqlCourseEnrollByCountryWorkflowJob job parameters.  Format as YAML.</description>
-            </hudson.model.TextParameterDefinition>
+                <defaultValue>{{ ANALYTICS_SCHEDULE_INSERT_TO_MYSQL_COURSE_ENROLL_BY_COUNTRY_EXTRA_VARS | default('') }}</defaultValue>
```
Maybe use for to generate that? Unless we go for using "groovy script modules" where this might be as well done in groovy.

pomegranited replied:
@jbzdak Yes.. I'd love to be able to use groovy for all of this, instead of XML + a separate groovy DSL.  I've tried it before, but ran into java dependency issues, but will try again next week while doing the github authentication change.

jbzdak commented on playbooks/roles/jenkins_master/defaults/main.yml:
```diff
@@ -100,4 +100,4 @@ jenkins_debian_pkgs:
 
 # Extra packages need for a specific jenkins instance.
-JENKINS_EXTRA_PKGS: []
+JENKINS_EXTRA_PKGS: "{{ jenkins_extra_pkgs | default([]) }}"
```
I think it would be cleaner to just add package installation to ``analytics_jenkins``.

pomegranited replied:
@jbzdak Done with [26069db](https://github.com/edx/configuration/pull/2939/commits/26069db90bc275bc7e75abc9e9262cb4fb41a9a2).

jbzdak commented on playbooks/roles/jenkins_master/defaults/main.yml:
```diff
@@ -4,7 +4,7 @@ jenkins_group: "edx"
 jenkins_port: 8080
 
-jenkins_version: "1.638"
+jenkins_version: "1.656"
```
1. This role is used also by other packages, we need to check if this doesn't break anyone's workflow ;)
2. Jenkins version is also embedded in ``config.xml`` template it would be good to check if this there were no config format changes in between that versions and update the version in ``config.xml``

mulby replied:
I share the concern about changing this version. This role is used by a lot of different systems here internally. We would need to do quite a bit of regression testing and rope in some other teams if we need to make this kind of change.

pomegranited replied:
@jbzdak and @mulby Gotcha, fixed with [f566aa4](https://github.com/edx/configuration/pull/2939/commits/f566aa4b7c9cf7cc69f4f7ec1dfcbee8be058e5d).  [See below](https://github.com/edx/configuration/pull/2939#discussion_r59113478) for more details.

jbzdak commented:
Tested and jenkins works as advertised :+1: .

mulby commented on playbooks/roles/jenkins_analytics/templates/jenkins_ssh_config.j2:
```diff
@@ -0,0 +1,2 @@
+Host *
+ForwardAgent yes
```
Is this necessary?

remote-task should be doing this for you: https://github.com/edx/edx-analytics-pipeline/blob/master/edx/analytics/tasks/launchers/remote.py#L257

pomegranited replied:
@jbzdak and @mulby [f566aa4](https://github.com/edx/configuration/pull/2939/commits/f566aa4b7c9cf7cc69f4f7ec1dfcbee8be058e5d) reverts the Jenkins version update and the .ssh/config file addition. They were both things I tried to get the secure repo cloned on the EMR instance, but neither were necessary.  

The real culprit turned out to be a [minor bug in the edx-analytics-pipeline (PR #232)](https://github.com/edx/edx-analytics-pipeline/pull/232).  Could you please review that when you get a chance?

mulby replied:
Looks good to me.

Should probably be reviewed by @jibsheet while Ed is away.
On Apr 9, 2016 6:29 AM, "Jillian Vogel" <notifications@github.com> wrote:

> In playbooks/roles/jenkins_analytics/templates/jenkins_ssh_config.j2
> <https://github.com/edx/configuration/pull/2939#discussion_r59113478>:
>
> > @@ -0,0 +1,2 @@
> > +Host *
> > +ForwardAgent yes
>
> @jbzdak <https://github.com/jbzdak> and @mulby <https://github.com/mulby>
> f566aa4
> <https://github.com/edx/configuration/pull/2939/commits/f566aa4b7c9cf7cc69f4f7ec1dfcbee8be058e5d>
> reverts the Jenkins version update and the .ssh/config file addition. They
> were both things I tried to get the secure repo cloned on the EMR instance,
> but neither were necessary.
>
> The real culprit turned out to be a minor bug in the
> edx-analytics-pipeline (PR #232)
> <https://github.com/edx/edx-analytics-pipeline/pull/232>. Could you
> please review that when you get a chance?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/configuration/pull/2939/files/f4bb297e4e6159d79f22a65208e9a1069762ec55#r59113478>
>

pomegranited commented:
@jbzdak Do you mind re-testing this change?  If you're using my analytics-sandbox-private repo branch, then testing this will also test [PR 232 on edx-analytics-pipeline](https://github.com/edx/edx-analytics-pipeline/pull/232).

jbzdak commented:
@pomegranited From the code standpoint this LGTM, the good thing is that we reduced number of moving parts we need to get right. 

It would be good to clarify steps between: "I have an root aws account" and "I have a working analytics sandbox with DSL Jenkins configured". While documenting setting up the sandbox is out of scope for this task, I believe that documenting how to fit and configure Jenkins instance would be important. 

I can test this documentation, but this shouldn't block merging this (as setting up analytics stack takes quite a long time).

jibsheet commented:
While working on this branch, I noticed that https://github.com/edx/configuration/pull/2928 isn't in this PR.  No need to cherry pick it in, but can you make sure any future PRs (like the github auth one) have been rebased far enough forward for it?

jibsheet commented:
I attempted to deploy a jenkins using this branch today (we're currently testing on a deploy from master, so we know the base works).

When trying to specify:

ANALYTICS_SCHEDULE_COURSE_ACTIVITY_WEEKLY_EXTRA_VARS: "@{{ ANALYTICS_SCHEDULE_SECURE_REPO_DEST }}/job_configs/course_activity.yml"

where course_activity.yml is

---
AUTOMATION_KEYPAIR_NAME: keyname
EMR_IAM_ROLE: EMR_EC2_DefaultRole
NUM_REDUCE_TASKS: '1'
NUM_TASK_INSTANCES: '1'
... more things elided because internal stuff

we get

ERROR: (AnalyticsTasks.groovy, line 375) No signature of method: java.lang.String.collectEntries() is applicable for argument types: (org.edx.jenkins.dsl.AnalyticsTasks$_parseYaml_closure1) values:
[org.edx.jenkins.dsl.AnalyticsTasks$_parseYaml_closure1@45a0ae7]

However, we even get this error without specifying any EXTRA_VARS for COURSE_ACTIVITY_WEEKLY. (I tested on a clean EC2 instance)

Can you check on the source of the error, but also confirm our format for the EXTRA_VARS file going forward?

Thanks

pomegranited commented:
@jibsheet I've rebased this PR to include the changes made to upstream/master.

I've addressed the bug you're seeing with an unspecified `<ANALYTICS_TASK>_EXTRA_VARS` with [e8bc878](https://github.com/edx/configuration/pull/2939/commits/e8bc878ad5660c62d3ffb1fb5aeabb14cf1c0481), by setting the default value for those `<ANALYTICS_TASK>_EXTRA_VARS` to `''` instead of `null`.  The `null` value was being translated by the seed_job_template.xml into the string `'None'`, which my AnalyticsTasks.parseYaml didn't handle well, resulting in the `java.lang.String.collectEntries()` error you saw.

So I also created [edx-jenkins-job-dsls PR #23](https://github.com/edx-ops/edx-jenkins-job-dsl/pull/23), which [addresses the `java.lang.String.collectEntries()` issue](https://github.com/edx-ops/edx-jenkins-job-dsl/pull/23/files#diff-42df318bb245d44fc2a494164cde3123L370), and adds unit tests to AnalyticsTasks.parseYaml to verify.  [As shown by the unit test](https://github.com/edx-ops/edx-jenkins-job-dsl/pull/23/files#diff-58d684bac582780f6d79555d60dc1f3bR20), `parseYaml('None')` returns an empty map `[:]`.

However, I was unable to reproduce the issue you were seeing with the example `course_activity.yml` file you gave?  See analytics sandbox [AnalyticsSeedJob run #62](http://52.20.136.133:8080/job/AnalyticsSeedJob/62/console), which resulted in a [CourseActivity task with these build parameters](http://52.20.136.133:8080/job/CourseActivityWeeklyTask/build?delay=0sec). Could you provide more of that example file, with a full list of the variables used, even if the values themselves need to be elided?

jibsheet commented:
@pomegranited My guess is that the file error was masked by the None settings (since we were only specifying one EXTRA_VARS file during this test run).  Rerunning this morning, it looks good, and I've done a quick confirmation that Gabe's overrides are being picked up.

When he is out of meetings, I'll have him verify the job and try running and he'll know for sure if the extra vars were picked up properly

jibsheet commented:
@pomegranited Our tests look good 👍 go ahead and merge.

jibsheet commented:
@pomegranited Actually, we noticed an issue as we got further into testing.
We set JENKINS_ANALYTICS_GITHUB_KEY and ANALYTICS_SCHEDULE_MASTER_SSH_CREDENTIAL_FILE to strings of the SSH key.

We don't want to put the EMR deploy private key in the ANALYTICS_SCHEDULE_SECURE_REPO_URL repo, it lives in a separate secure repo which is not checked out as part of the run on the jenkins box.

Essentially, I run ansible-playbook  -i 10.x.x.x, ../run_role.yml -e 'role=jenkins_analytics'  -e@/path/to/secure/vars/file.yml 

I don't think I can use a lookup() here since the job wants to run on the jenkins box, but is there anything else I can do to not have to store the file in the ANALYTICS_SCHEDULE_SECURE_REPO_URL repo?

If not, we'll cut more keys and put them over there, but we were going to avoid that complexity.

pomegranited commented:
@jibsheet You can provide an explicit private key instead of pointing to a file, but you have to construct your own `JENKINS_ANALYTICS_CREDENTIALS ` list, because the default value uses `type: ssh-private-keyfile` for the ssh credential.  I've [updated the README to illustrate how to do this](7c370ea133853655c65c4fb7a2ffbff64c03885e).

Still ok to merge?

jibsheet commented:
@pomegranited Cool - I thought about that on the commute home and wondered if it would be the right solution.  I should be able to implement and test that out tomorrow.  Feel free to merge when travis is done.
